### ISSUE TITLE
Test failure due to correlated subquery in IN predicate

### ISSUE THREAD

**Issue by martint** [prestodb/presto#8165]
TestRedisDistributedHash.testCorrelatedInPredicateSubqueries fails with the following error:

```
java.lang.AssertionError: Execution of 'actual' query failed: SELECT orderkey, clerk IN (SELECT clerk FROM orders s WHERE s.custkey = o.custkey AND s.orderkey < o.orderkey) FROM orders o

	at org.testng.Assert.fail(Assert.java:83)
	at com.facebook.presto.tests.QueryAssertions.assertQuery(QueryAssertions.java:86)
	at com.facebook.presto.tests.AbstractTestQueryFramework.assertQuery(AbstractTestQueryFramework.java:117)
	at com.facebook.presto.tests.AbstractTestQueryFramework.assertQuery(AbstractTestQueryFramework.java:112)
	at com.facebook.presto.tests.AbstractTestQueries.testCorrelatedInPredicateSubqueries(AbstractTestQueries.java:7166)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.testng.internal.MethodInvocationHelper.invokeMethod(MethodInvocationHelper.java:104)
	at org.testng.internal.Invoker.invokeMethod(Invoker.java:645)
	at org.testng.internal.Invoker.invokeTestMethod(Invoker.java:851)
	at org.testng.internal.Invoker.invokeTestMethods(Invoker.java:1177)
	at org.testng.internal.TestMethodWorker.invokeTestMethods(TestMethodWorker.java:129)
	at org.testng.internal.TestMethodWorker.run(TestMethodWorker.java:112)
	at org.testng.TestRunner.privateRun(TestRunner.java:756)
	at org.testng.TestRunner.run(TestRunner.java:610)
	at org.testng.SuiteRunner.runTest(SuiteRunner.java:387)
	at org.testng.SuiteRunner.runSequentially(SuiteRunner.java:382)
	at org.testng.SuiteRunner.privateRun(SuiteRunner.java:340)
	at org.testng.SuiteRunner.run(SuiteRunner.java:289)
	at org.testng.SuiteRunnerWorker.runSuite(SuiteRunnerWorker.java:52)
	at org.testng.SuiteRunnerWorker.run(SuiteRunnerWorker.java:86)
	at org.testng.TestNG.runSuitesSequentially(TestNG.java:1293)
	at org.testng.TestNG.runSuitesLocally(TestNG.java:1218)
	at org.testng.TestNG.runSuites(TestNG.java:1133)
	at org.testng.TestNG.run(TestNG.java:1104)
	at org.testng.IDEARemoteTestNG.run(IDEARemoteTestNG.java:72)
	at org.testng.RemoteTestNGStarter.main(RemoteTestNGStarter.java:127)
Caused by: java.lang.IllegalStateException: No mapping for column handle RedisColumnHandle{connectorId=redis, ordinalPosition=0, name=redis_key, type=varchar(64), mapping=null, dataFormat=null, formatHint=null, keyDecoder=true, hidden=true, internal=false}!
	at com.google.common.base.Preconditions.checkState(Preconditions.java:518)
	at com.facebook.presto.redis.decoder.hash.HashRedisRowDecoder.decodeRow(HashRedisRowDecoder.java:58)
	at com.facebook.presto.redis.RedisRecordCursor.nextRow(RedisRecordCursor.java:184)
	at com.facebook.presto.redis.RedisRecordCursor.advanceNextPosition(RedisRecordCursor.java:147)
	at com.facebook.presto.spi.RecordPageSource.getNextPage(RecordPageSource.java:99)
	at com.facebook.presto.operator.TableScanOperator.getOutput(TableScanOperator.java:262)
	at com.facebook.presto.operator.Driver.processInternal(Driver.java:303)
	at com.facebook.presto.operator.Driver.lambda$processFor$6(Driver.java:234)
	at com.facebook.presto.operator.Driver.tryWithLock(Driver.java:537)
	at com.facebook.presto.operator.Driver.processFor(Driver.java:229)
	at com.facebook.presto.execution.SqlTaskExecution$DriverSplitRunner.processFor(SqlTaskExecution.java:623)
	at com.facebook.presto.execution.executor.PrioritizedSplitRunner.process(PrioritizedSplitRunner.java:162)
	at com.facebook.presto.execution.executor.TaskExecutor$TaskRunner.run(TaskExecutor.java:463)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
```

This seems to be a regression introduced in 98faf69c110b0f7e51b79924fa06aef9ae345a6e

**Comment by martint** [comment]
I believe this is a bug in the redis connector. What seems to be happening is that the connector initially presents a set of internal columns to the engine during planning that it can't produce data for during execution. Because of the way IN predicates with correlated subqueries are currently implemented, those columns cannot be pruned from the output of the tablescan. This is what the plan looks like:

```
- Output[orderkey, _col1] => [orderkey:bigint, expr_21:boolean]
         Cost: {rows: ?, bytes: ?}
         _col1 := expr_21
     - RemoteExchange[GATHER] => expr_21:boolean, orderkey:bigint
             Cost: {rows: ?, bytes: ?}
         - Project[] => [expr_21:boolean, orderkey:bigint]
                 Cost: {rows: ?, bytes: ?}
                 expr_21 := (CASE WHEN ("countmatches" > BIGINT '0') THEN true WHEN ("countnullmatches" > BIGINT '0') THEN CAST(null AS boolean) ELSE false END)
             - Project[] => [orderkey:bigint, countnullmatches:bigint, countmatches:bigint]
                     Cost: {rows: ?, bytes: ?}
                 - Aggregate(FINAL)[redis_key, orderkey, custkey, orderstatus, totalprice, orderdate, orderpriority, clerk, shippriority, comment, _key, _value, _key_length, _value_length, _key_corrupt, _value_corrupt, unique][$hashvalue] => [redis_key:varc
                         Cost: {rows: ?, bytes: ?}
                         countnullmatches := "count"("count_34")
                         countmatches := "count"("count")
                     - LocalExchange[HASH][$hashvalue] ("redis_key", "orderkey", "custkey", "orderstatus", "totalprice", "orderdate", "orderpriority", "clerk", "shippriority", "comment", "_key", "_value", "_key_length", "_value_length", "_key_corrupt", "_va
                             Cost: {rows: ?, bytes: ?}
                         - Aggregate(PARTIAL)[redis_key, orderkey, custkey, orderstatus, totalprice, orderdate, orderpriority, clerk, shippriority, comment, _key, _value, _key_length, _value_length, _key_corrupt, _value_corrupt, unique][$hashvalue_40] => [r
                                 Cost: {rows: ?, bytes: ?}
                                 count_34 := "count"(*) (mask = expr_33)
                                 count := "count"(*) (mask = expr_32)
                             - Project[] => [expr_32:boolean, expr_33:boolean, redis_key:varchar(64), orderkey:bigint, custkey:bigint, orderstatus:varchar(1), totalprice:double, orderdate:date, orderpriority:varchar(15), clerk:varchar(15), shippriority:inte
                                     Cost: {rows: ?, bytes: ?}
                                     expr_32 := (("orderstatus" IS NOT NULL) AND ("orderstatus_4" IS NOT NULL))
                                     expr_33 := (("buildsideknownnonnull" IS NOT NULL) AND ((NOT ("orderstatus" IS NOT NULL)) OR (NOT ("orderstatus_4" IS NOT NULL))))
                                     $hashvalue_40 := "combine_hash"("combine_hash"("combine_hash"("combine_hash"("combine_hash"("combine_hash"("combine_hash"("combine_hash"("combine_hash"("combine_hash"("combine_hash"("combine_hash"("combine_hash"("combine
                                 - LeftJoin[("custkey" = "custkey_3") AND (((("orderstatus" IS NULL) OR ("orderstatus" = "orderstatus_4")) OR ("orderstatus_4" IS NULL)) AND ("orderkey_2" < "orderkey"))][$hashvalue_35, $hashvalue_37] => [redis_key:varchar(64
                                         Cost: {rows: ?, bytes: ?}
                                     - RemoteExchange[REPARTITION][$hashvalue_35] => redis_key:varchar(64), orderkey:bigint, custkey:bigint, orderstatus:varchar(1), totalprice:double, orderdate:date, orderpriority:varchar(15), clerk:varchar(15), shippriorit
                                             Cost: {rows: ?, bytes: ?}
                                         - Project[] => [redis_key:varchar(64), orderkey:bigint, custkey:bigint, orderstatus:varchar(1), totalprice:double, orderdate:date, orderpriority:varchar(15), clerk:varchar(15), shippriority:integer, comment:varchar(7
                                                 Cost: {rows: ?, bytes: ?}
                                                 $hashvalue_36 := "combine_hash"(BIGINT '0', COALESCE("$operator$hash_code"("custkey"), 0))
                                             - AssignUniqueId => [redis_key:varchar(64), orderkey:bigint, custkey:bigint, orderstatus:varchar(1), totalprice:double, orderdate:date, orderpriority:varchar(15), clerk:varchar(15), shippriority:integer, comment:
                                                     Cost: {rows: ?, bytes: ?}
                                                 - TableScan[redis:RedisTableHandle{connectorId=redis, schemaName=tpch, tableName=orders, keyDataFormat=raw, valueDataFormat=hash, keyName=null}, originalConstraint = true] => [redis_key:varchar(64), orderkey:
                                                         Cost: {rows: ?, bytes: ?}
                                                         redis_key := RedisColumnHandle{connectorId=redis, ordinalPosition=0, name=redis_key, type=varchar(64), mapping=null, dataFormat=null, formatHint=null, keyDecoder=true, hidden=true, internal=false}
                                                         orderkey := RedisColumnHandle{connectorId=redis, ordinalPosition=1, name=orderkey, type=bigint, mapping=orderkey, dataFormat=null, formatHint=null, keyDecoder=false, hidden=false, internal=false}
                                                         custkey := RedisColumnHandle{connectorId=redis, ordinalPosition=2, name=custkey, type=bigint, mapping=custkey, dataFormat=null, formatHint=null, keyDecoder=false, hidden=false, internal=false}
                                                         orderstatus := RedisColumnHandle{connectorId=redis, ordinalPosition=3, name=orderstatus, type=varchar(1), mapping=orderstatus, dataFormat=null, formatHint=null, keyDecoder=false, hidden=false, interna
                                                         totalprice := RedisColumnHandle{connectorId=redis, ordinalPosition=4, name=totalprice, type=double, mapping=totalprice, dataFormat=null, formatHint=null, keyDecoder=false, hidden=false, internal=false
                                                         orderdate := RedisColumnHandle{connectorId=redis, ordinalPosition=5, name=orderdate, type=date, mapping=orderdate, dataFormat=iso8601, formatHint=null, keyDecoder=false, hidden=false, internal=false}
                                                         orderpriority := RedisColumnHandle{connectorId=redis, ordinalPosition=6, name=orderpriority, type=varchar(15), mapping=orderpriority, dataFormat=null, formatHint=null, keyDecoder=false, hidden=false,
                                                         clerk := RedisColumnHandle{connectorId=redis, ordinalPosition=7, name=clerk, type=varchar(15), mapping=clerk, dataFormat=null, formatHint=null, keyDecoder=false, hidden=false, internal=false}
                                                         shippriority := RedisColumnHandle{connectorId=redis, ordinalPosition=8, name=shippriority, type=integer, mapping=shippriority, dataFormat=null, formatHint=null, keyDecoder=false, hidden=false, interna
                                                         comment := RedisColumnHandle{connectorId=redis, ordinalPosition=9, name=comment, type=varchar(79), mapping=comment, dataFormat=null, formatHint=null, keyDecoder=false, hidden=false, internal=false}
                                                         _key := RedisColumnHandle{connectorId=redis, ordinalPosition=10, name=_key, type=varchar, mapping=null, dataFormat=null, formatHint=null, keyDecoder=false, hidden=true, internal=true}
                                                         _value := RedisColumnHandle{connectorId=redis, ordinalPosition=11, name=_value, type=varchar, mapping=null, dataFormat=null, formatHint=null, keyDecoder=false, hidden=true, internal=true}
                                                         _key_length := RedisColumnHandle{connectorId=redis, ordinalPosition=12, name=_key_length, type=bigint, mapping=null, dataFormat=null, formatHint=null, keyDecoder=false, hidden=true, internal=true}
                                                         _value_length := RedisColumnHandle{connectorId=redis, ordinalPosition=13, name=_value_length, type=bigint, mapping=null, dataFormat=null, formatHint=null, keyDecoder=false, hidden=true, internal=true}
                                                         _key_corrupt := RedisColumnHandle{connectorId=redis, ordinalPosition=14, name=_key_corrupt, type=boolean, mapping=null, dataFormat=null, formatHint=null, keyDecoder=false, hidden=true, internal=true}
                                                         _value_corrupt := RedisColumnHandle{connectorId=redis, ordinalPosition=15, name=_value_corrupt, type=boolean, mapping=null, dataFormat=null, formatHint=null, keyDecoder=false, hidden=true, internal=tr
                                     - LocalExchange[HASH][$hashvalue_37] ("custkey_3") => orderstatus_4:varchar(1), orderkey_2:bigint, custkey_3:bigint, buildsideknownnonnull:bigint, $hashvalue_37:bigint
                                             Cost: {rows: ?, bytes: ?}
                                          ...
```

For now, I'm going to disable those tests for the redis connector. Separately, we need to evaluate whether the plan that's being generated is the best we can do (in particular, whether we need *all* the columns from the table given that we're attaching a unique id to every row).

**Comment by findepi** [comment]
@martint actually redis connector has problem producing `redis_key` column which is not an "intrinsic" column. In our tests this is the column mapped to redis key and it's defined as hidden. I updated your comment to point that column too.

>  Separately, we need to evaluate whether the plan that's being generated is the best we can do (in particular, whether we need all the columns from the table given that we're attaching a unique id to every row).

Surely we can do better. 

First off, in the plan there is:
```
Aggregate[redis_key, orderkey, custkey, orderstatus, totalprice, orderdate, orderpriority,
clerk, shippriority, comment, _key, _value, _key_length, _value_length, _key_corrupt,
_value_corrupt, unique]
                         countnullmatches := "count"("count_34")
                         countmatches := "count"("count")
```
after #8168 this will be:
```
Aggregate[orderkey, comment, unique]
                         countnullmatches := "count"("count_34")
                         countmatches := "count"("count")
```
generally, this is an aggregation over _all_ fields that we use later on. Since `unique` is a generated column that is unique before Left Outer Join with de-correlated IN's build side, the aggregation is as "aggregative" as the one over just `unique` column would be. Additional columns appear in aggregation key for _technical_ reason so they appear in its output.

If we had an aggregation function like `single_value(..)` that returns single value  (and fails if there are 2+ values to aggregate) we could use that instead. Quite simple and doesn't introduce much complexity, I believe.

Second, we do the counts (very same aggregation), but are interested only whether result was 0 or not. This could be more efficiently served with a dedicated operator that does count with limit. This introduces more complexity (new operator(s)) and should be deferred until we see real-world cases that this work would greatly improve.
